File: backend/api/schemas/dashboards.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from persistence.models import Dashboard, DashboardWidget
from visualization.selection import compatible_chart_types

ChartType = Literal["table", "kpi", "bar", "line", "pie", "area"]


class DashboardCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=1_000)

    @field_validator("name")
    @classmethod
    def strip_name(cls, value: str) -> str:
        if not (clean := value.strip()):
            raise ValueError("Dashboard name must not be blank")
        return clean


class DashboardSummary(BaseModel):
    id: UUID
    name: str
    description: str | None
    widget_count: int
    created_at: datetime
    updated_at: datetime


class DashboardWidgetCreate(BaseModel):
    query_run_id: UUID | None = None
    saved_analysis_id: UUID | None = None
    title: str = Field(min_length=1, max_length=160)
    chart_type: ChartType

    @field_validator("title")
    @classmethod
    def strip_title(cls, value: str) -> str:
        if not (clean := value.strip()):
            raise ValueError("Widget title must not be blank")
        return clean

    @model_validator(mode="after")
    def require_source(self) -> "DashboardWidgetCreate":
        if (self.query_run_id is None) == (self.saved_analysis_id is None):
            raise ValueError("Provide exactly one query run or saved analysis")
        return self


class DashboardWidgetUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=160)
    chart_type: ChartType | None = None
    position: int | None = Field(default=None, ge=0)

    @field_validator("title")
    @classmethod
    def strip_optional_title(cls, value: str | None) -> str | None:
        if value is None:
            return None
        if not (clean := value.strip()):
            raise ValueError("Widget title must not be blank")
        return clean


class DashboardWidgetResponse(BaseModel):
    id: UUID
    dashboard_id: UUID
    datasource_id: UUID
    source_query_run_id: UUID | None
    title: str
    question: str
    validated_query: dict[str, Any]
    query_type: str
    chart_type: str
    chart_config: dict[str, Any]
    compatible_chart_types: list[str]
    position: int
    result: dict[str, Any] | None
    status: str
    row_count: int | None
    duration_ms: int | None
    last_refreshed_at: datetime | None
    error: dict[str, str] | None
    created_at: datetime
    updated_at: datetime


class DashboardDetail(DashboardSummary):
    widgets: list[DashboardWidgetResponse]


def widget_response(widget: DashboardWidget) -> DashboardWidgetResponse:
    return DashboardWidgetResponse(
        id=widget.id,
        dashboard_id=widget.dashboard_id,
        datasource_id=widget.datasource_id,
        source_query_run_id=widget.source_query_run_id,
        title=widget.title,
        question=widget.question,
        validated_query=widget.validated_query,
        query_type=widget.query_type,
        chart_type=widget.chart_type,
        chart_config=widget.chart_config,
        compatible_chart_types=compatible_chart_types(widget.result_json),
        position=widget.position,
        result=widget.result_json,
        status=widget.status,
        row_count=widget.row_count,
        duration_ms=widget.duration_ms,
        last_refreshed_at=widget.last_refreshed_at,
        error=(
            {
                "code": widget.last_error_code,
                "message": widget.last_error_message or "Refresh failed",
            }
            if widget.last_error_code
            else None
        ),
        created_at=widget.created_at,
        updated_at=widget.updated_at,
    )


def dashboard_summary(dashboard: Dashboard, widget_count: int) -> DashboardSummary:
    return DashboardSummary(
        id=dashboard.id,
        name=dashboard.name,
        description=dashboard.description,
        widget_count=widget_count,
        created_at=dashboard.created_at,
        updated_at=dashboard.updated_at,
    )

File: frontend/lib/dashboards.ts
import { apiRequest } from "@/lib/api-client";
import type { QueryResult } from "@/lib/query-runs";
import type { ChartType } from "@/lib/visualization";

export interface DashboardSummary {
  id: string;
  name: string;
  description: string | null;
  widget_count: number;
  created_at: string;
  updated_at: string;
}

export interface DashboardWidget {
  id: string;
  dashboard_id: string;
  datasource_id: string;
  source_query_run_id: string | null;
  title: string;
  question: string;
  validated_query: { sql: string; expected_columns: string[] };
  query_type: string;
  chart_type: ChartType;
  chart_config: Record<string, unknown>;
  compatible_chart_types: ChartType[];
  position: number;
  result: QueryResult | null;
  status: "ready" | "empty" | "stale" | "failed";
  row_count: number | null;
  duration_ms: number | null;
  last_refreshed_at: string | null;
  error: { code: string; message: string } | null;
  created_at: string;
  updated_at: string;
}

export interface DashboardDetail extends DashboardSummary {
  widgets: DashboardWidget[];
}

export const listDashboards = () => apiRequest<DashboardSummary[]>("/api/v1/dashboards");

export const createDashboard = (name: string, description?: string) =>
  apiRequest<DashboardSummary>("/api/v1/dashboards", {
    method: "POST",
    body: JSON.stringify({ name, description: description || null }),
  });

export const getDashboard = (id: string) =>
  apiRequest<DashboardDetail>(`/api/v1/dashboards/${id}`);

export const addDashboardWidget = (
  dashboardId: string,
  payload: {
    query_run_id?: string;
    saved_analysis_id?: string;
    title: string;
    chart_type: ChartType;
  },
) =>
  apiRequest<DashboardWidget>(`/api/v1/dashboards/${dashboardId}/widgets`, {
    method: "POST",
    body: JSON.stringify(payload),
  });

export const updateDashboardWidget = (
  widgetId: string,
  payload: { title?: string; chart_type?: ChartType; position?: number },
) =>
  apiRequest<DashboardWidget>(`/api/v1/dashboard-widgets/${widgetId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });

export const deleteDashboardWidget = (widgetId: string) =>
  apiRequest<void>(`/api/v1/dashboard-widgets/${widgetId}`, { method: "DELETE" });

export const refreshDashboardWidget = (widgetId: string) =>
  apiRequest<DashboardWidget>(`/api/v1/dashboard-widgets/${widgetId}/refresh`, {
    method: "POST",
  });
